src/pipes/erldns_questions.erl

```diff
@@ -0,0 +1,32 @@
+-module(erldns_questions).
+-moduledoc """
+Remove all redundant questions from a DNS message,
+and parses the first question into a list of labels.
+
+## Telemetry events
+
+- `[erldns, pipeline, questions]` with `#{count => non_neg_integer()}`
+where `count` is the number of questions removed.
+""".
+
+-include_lib("dns_erlang/include/dns.hrl").
+
+-behaviour(erldns_pipeline).
+
+-export([call/2]).
+
+-doc "`c:erldns_pipeline:call/2` callback.".
+-spec call(dns:message(), erldns_pipeline:opts()) -> erldns_pipeline:return().
+call(#dns_message{questions = []} = Msg, _) ->
+    {stop, Msg#dns_message{qr = true}};
+call(#dns_message{questions = [#dns_query{} = Q1]} = Msg, Opts) ->
+    Labels = dns:dname_to_lower_labels(Q1#dns_query.name),
+    {Msg, Opts#{query_labels := Labels, query_type := Q1#dns_query.type}};
+call(#dns_message{questions = [#dns_query{} = Q1, Q2 | Rest]} = Msg, #{host := Host} = Opts) ->
+    Labels = dns:dname_to_lower_labels(Q1#dns_query.name),
+    Measurements = #{count => 1 + length(Rest)},
+    Metadata = #{host => Host, questions => [Q1, Q2 | Rest]},
+    telemetry:execute([erldns, pipeline, questions], Measurements, Metadata),
```

Comment:
Multi-question messages are not of great interest to us, since in practice we only support a single question, and any additional questions can be ignored safely.